[InstCombine] Factorise add/sub and max/min using distributivity

jf-botto opened the pull request:
This PR fixes part of [Issue 92433](https://github.com/llvm/llvm-project/issues/92433).

The alive proof sometimes times out. I have to add `noundef` to certain variables to make it not time out more consistently. Given how min/max/add preserve undef, I believe these optimisations to be correct when undef values are passed through.


Proofs:
1. https://alive2.llvm.org/ce/z/9N_cAE (1 - 4 | umin_of_umax, umin_of_umax_comm, smin_of_smax, smin_of_smax_comm)
2. https://alive2.llvm.org/ce/z/DchqXW (5 - 8 | umax_of_umin, umax_of_umin_comm, smax_of_smin, smax_of_smin_comm)
3. https://alive2.llvm.org/ce/z/2j3P-- (9 - 10 | umax_of_uadd_sat, umax_of_uadd_sat_comm)
4. https://alive2.llvm.org/ce/z/vR4v7c (11 | umin_of_uadd_sat)
5. https://alive2.llvm.org/ce/z/dQjJT4 (12 | umin_of_uadd_sat_comm)
6. https://alive2.llvm.org/ce/z/MskzQc (13 | smax_of_sadd_sat)
7. https://alive2.llvm.org/ce/z/kLys_n (14 | smax_of_sadd_sat_comm)
8. https://alive2.llvm.org/ce/z/b2KQeS (15 | smin_of_sadd_sat)
9. https://alive2.llvm.org/ce/z/R4JrTN (16 | smin_of_sadd_sat_comm)

nikic commented:
> [](https://alive2.llvm.org/ce/z/ZaXf7L)
> 
> The alive proof sometimes times out. I have to add `noundef` to certain variables to make it not time out more consistently. Given how min/max/add preserve undef, I believe these optimisations to be correct when undef values are passed through.

Try replacing all the i32s with i8. That usually helps with timeouts.

goldsteinn commented:
> > The alive proof sometimes times out. I have to add `noundef` to certain variables to make it not time out more consistently. Given how min/max/add preserve undef, I believe these optimisations to be correct when undef values are passed through.
> 
> Try replacing all the i32s with i8. That usually helps with timeouts.

In this case even `i4` times out: https://alive2.llvm.org/ce/z/UPjvkJ

jf-botto commented:
> > The alive proof sometimes times out. I have to add `noundef` to certain variables to make it not time out more consistently. Given how min/max/add preserve undef, I believe these optimisations to be correct when undef values are passed through.
> 
> Try replacing all the i32s with i8. That usually helps with timeouts.

Thank you! That really helps. I will update the alive link. I may have to separate the tests into 3-4 links.

goldsteinn commented:
Two of the proofs are buggy: https://alive2.llvm.org/ce/z/Xz2-Vq (`src5` -> `tgt5` and `src6` -> `tgt6`).

jf-botto commented:
> > > The alive proof sometimes times out. I have to add `noundef` to certain variables to make it not time out more consistently. Given how min/max/add preserve undef, I believe these optimisations to be correct when undef values are passed through.
> > 
> > 
> > Try replacing all the i32s with i8. That usually helps with timeouts.
> 
> In this case even `i4` times out: https://alive2.llvm.org/ce/z/UPjvkJ

It works for individual tests.

> Two of the proofs are buggy: https://alive2.llvm.org/ce/z/Xz2-Vq (`src5` -> `tgt5` and `src6` -> `tgt6`).

Thank you. Just noticed that. Will fix it.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1751,9 +1830,18 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
+  case Intrinsic::smax: {
+    if (Instruction *I = foldCallUsingDistributiveLaws(II, Builder))
+      return I;
+
```
Don't think you need the call in the `umin`/`umax`/`smax` cases, they all fallthrough to `smin` (and it seems a bit silly to potentially fail 4x times on the same call).

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1505,6 +1505,80 @@ foldMinimumOverTrailingOrLeadingZeroCount(Value *I0, Value *I1,
+// Attempts to factorise a common term
+// in an instruction that has the form "(A op' B) op (C op' D)
+static Instruction *
+foldCallUsingDistributiveLaws(CallInst *II, InstCombiner::BuilderTy &Builder) {
```
`CallInst` -> `IntrinsicInst`

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1505,6 +1505,80 @@ foldMinimumOverTrailingOrLeadingZeroCount(Value *I0, Value *I1,
+
+  if (LHS && RHS) {
+    CallInst *Op0 = dyn_cast<CallInst>(LHS);
+    CallInst *Op1 = dyn_cast<CallInst>(RHS);
```
Again, `IntrinsicInst`.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1505,6 +1505,80 @@ foldMinimumOverTrailingOrLeadingZeroCount(Value *I0, Value *I1,
+static Instruction *
+foldCallUsingDistributiveLaws(CallInst *II, InstCombiner::BuilderTy &Builder) {
+  Value *LHS = II->getOperand(0), *RHS = II->getOperand(1);
+  Intrinsic::ID TopLevelOpcode = II->getCalledFunction()->getIntrinsicID();
```
Assume you use `IntrinsicInst`, this can be `II->getIntrinsicID()`

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1505,6 +1505,80 @@ foldMinimumOverTrailingOrLeadingZeroCount(Value *I0, Value *I1,
+      Value *NewIntrinsic = Builder.CreateBinaryIntrinsic(TopLevelOpcode, B, D);
+      Function *F = Intrinsic::getDeclaration(II->getModule(), InnerOpcode,
+                                              II->getType());
+      return CallInst::Create(F, {NewIntrinsic, A});
```
This can be:
```
return replaceInstUsesWith(*II, Builder.CreateBinaryIntrinsic(TopLevelOpcode, B, D)); 
```

Similary below.

goldsteinn commented on llvm/test/Transforms/InstCombine/intrinsic-distributive.ll:
```diff
@@ -0,0 +1,210 @@
+;
+  %add1 = call i32 @llvm.sadd.sat.i32(i32 %z, i32 %x)
+  %add2 = call i32 @llvm.sadd.sat.i32(i32 %z, i32 %y)
+  %min = call i32 @llvm.smin.i32(i32 %add1, i32 %add2)
```
Can you change these tests to `i8` so its easier to verify them?

jf-botto committed:
Precommit test

jf-botto committed:
Adding missed optimisation

jf-botto commented:
Thank you @goldsteinn for the time and effort you've put into reviewing my code. I've addressed all of your points. 
The two test cases that were buggy, was simply a mistake I made when copying and adapting the tests into alive. I've corrected the Alive tests in the top comment.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1505,6 +1505,78 @@ foldMinimumOverTrailingOrLeadingZeroCount(Value *I0, Value *I1,
+  Value *LHS = II->getOperand(0), *RHS = II->getOperand(1);
+  Intrinsic::ID TopLevelOpcode = II->getIntrinsicID();
+
+  if (LHS && RHS) {
```
```suggestion
```
They are always valid.

jf-botto replied:
Added

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1505,6 +1505,78 @@ foldMinimumOverTrailingOrLeadingZeroCount(Value *I0, Value *I1,
+    if (!Op0 || !Op1)
+      return nullptr;
+
+    if (Op0->getIntrinsicID() != Op1->getIntrinsicID())
```
```suggestion
    if (!Op0->hasOneUse() || !Op1->hasOneUse())
      return nullptr;
    if (Op0->getIntrinsicID() != Op1->getIntrinsicID())
```

We need one-use check here to avoid creating more instructions than before :)
See also https://llvm.org/docs/InstCombineContributorGuide.html#multi-use-handling.

jf-botto replied:
I've added the `hasOneUse`. Did you mean to remove the intrinsic check @dtcxzyw?

goldsteinn replied:
He moved them below, what you have now is fine.

dtcxzyw replied:
Sorry, ignore me...

I will take a deeper look and provide a reproducer later.

dtcxzyw commented:
With one-use check, this patch doesn't change anything :(

I think we should handle the interesting cases I listed in https://github.com/llvm/llvm-project/issues/92433#issuecomment-2117064459 first.

dtcxzyw commented:
> @dtcxzyw, I was planning on making a second PR built upon this to handle the cases you mention. Would you rather me use this PR _just_ for those cases you mention?

Please file a separate PR.
